lib/middleware/operation.js

```diff
@@ -32,55 +36,95 @@ function findPropertyOperations ({ types, method, term, resource }) {
   return operations
 }
 
-function factory (api) {
-  return async (req, res, next) => {
-    if (!req.hydra.resource) {
-      return next()
+function mapOperations ({ api, res, method }) {
+  return res.locals.hydra.resources.flatMap((resource) => {
+    let moreOperations
+    if ('property' in resource) {
+      moreOperations = findPropertyOperations({ resource, method, api }).toArray()
+    } else {
+      const types = clownface({ ...api, term: [...resource.types] })
+      moreOperations = findClassOperations(types, method).toArray()
+    }
+
+    if (moreOperations.length) {
+      return moreOperations.map(operation => ({ resource, operation }))
     }
 
+    return [{ resource, operation: null }]
+  })
+}
+
+function findOperation (operationMiddleware) {
+  return async function (req, res, next) {
     const method = req.method === 'HEAD' ? 'GET' : req.method
-    const types = clownface({ ...api, term: [...req.hydra.resource.types] })
 
-    let operations
-    if (req.hydra.term.equals(req.hydra.resource.term)) {
-      // only look for direct operation when the root resource is requested
-      operations = findClassOperations(types, method)
+    let operations = mapOperations({ api: req.hydra.api, res, method })
+      .filter(({ operation }) => operation)
+
+    if (operationMiddleware) {
+      operations = operationMiddleware(operations, { req, res })
+    }
+
+    if (!operations.length) {
+      return next()
     }
-    if (!operations) {
-      // otherwise try finding the operation by property usage
-      operations = findPropertyOperations({ types, method, term: req.hydra.term, resource: req.hydra.resource })
+
+    if (operations.length > 1) {
+      error('Multiple operations found')
+      return next(new Error('Ambiguous operation'))
     }
 
-    const [operation, ...rest] = (operations || []).toArray()
+    const [{ resource, operation }] = operations
+    req.hydra.resource = resource
+    req.hydra.operation = operation
+    return next()
+  }
+}
+
+async function invokeOperation (req, res, next) {
+  const { api } = req.hydra
+
+  if (!req.hydra.operation) {
+    const operationMap = mapOperations({ api, res })
```

Comment:
`mapOperations` does not call `operationMiddleware`, if given. For the edge case that an additional operation would be added with a different method, that method would not be included in the allowed list. I'm not sure if it needs to be covered. An additional operation would mean it was not discovered using the API description, so it's anyway somehow hidden. But it could be also the case, that we need it for inline operations, once we add that feature.

Reply:
The core benefit I'm looking for right now is ability to disambiguate operations if multiple are found.

While this also allows adding/replacing operations, I would say that one does on their own responsibility

Reply:
We can keep it like this now, but in a separate PR we should have a look at all the edge cases.